### DISCUSSION TITLE
Reworked SMS.ino

### DISCUSSION THREAD

**Discussion by PsyK0p4T** [sanni/cartreader#742] [Look what I made]
First, a big "thank you" to all contributors of the current SMS.ino file, for everything done so far. This project and people here are amazing!

With all the successive additions brought to the file, and the adapter number increasing, I wanted to rework the file a bit for myself. Added things, simplified others... and it finally evolved a lot. I'm sharing it here, and let you decide if you want to replace the current file with this one.

[SMS.zip](https://github.com/sanni/cartreader/files/11045204/SMS.zip)

Quick changelog : 
- many functions, variables, arrays renamed
- new constants are used for adapters and systems (used to simplify all conditions and checks)
- dropped the feature to change adapter mode between two dumps (hard to maintain and adds more conditions)
- added the Retron 3in1 adapter for SMS dumping
- added all possible rom sizes for SG-1000 and SMS/GG carts
- during the rom size manual selection, the offered values are adapted to the selected system
- moved code (function calls, function order...)
- comments added/modified

Has been tested successfully with : 
- hw3 and hw5
- SMS, SEGA Card, Mark-III, SG-1000, and GameGear cartridges
- SMS-to-MD raphnet, MIII-to-MD raphnet, Retrode SMS, Retrode GG, Retron 3in1 SMS, Retron 3in1 GG adapters

Known issues or behaviors (already present in current SMS.ino) : 
- SG-1000 : games with 40k mapping are not yet supported (40 and 48kb rom sizes are currently commented in code)
- GG : some games seem to be better recognized with Retron 3in1 adapter than with Retrode one
- GG : Codemasters mapper need to be added
- SMS : more testing needed with [memory paging chip](https://github.com/sanni/cartreader/issues/527) cartridges. Using a Retron 3in1 adapter for example.
- SMS : testers with Retrofreak 3in1 adapter would be appreciated. At least to confirm if it can be used exactly like the Retron 3in1.
- SMS/MIII/SG-1000 : some SEGA Card are not recognized

**Comment by Ancyker** [comment]
I tried it with monopoly just to see if anything changed...
```
OSCR HW5 V12.4

[+] SMS/GG/MIII/SG-1000

[+] SMS/Mark 3 raphnet

[+] Read Rom


SMS Header Info
 
Name: TMR SEGA
Size: 512KB
 
Press Button...

Saving to SMS/ROM/TMR SEGA/36/...
[*******************]
CRC32... B4DB4ACD -> Not found
```

Same issue. Appears to be garbage data after 0x1FFFF:
![image](https://user-images.githubusercontent.com/7105333/227133494-36638c37-e494-48f2-a807-b9c9d4bf4a80.png)

**Reply by PsyK0p4T** [reply]
Thanks for the logs. Mystery solved as discussed [here](https://github.com/sanni/cartreader/issues/527#issuecomment-1480861889) :)

**Reply by PsyK0p4T** [reply]
Code update a bit to solve warnings and adapt to latest commits ;) Will be pushed soon
